### PR TITLE
[FLINK-27767][sql-gateway] Introduce Endpoint API and utils

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Introdue Endpoint API and utils to create endpoint.*

## Brief change log

  - *Introduce the Endpoint API*
  - *Introduce the Factory to create the endpoint*
  - *Introduce the FactoryUtils to find the EndpointFactory.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added tests to create the Mocked endpoint.*
  - *Added negative tests to create the endpoint.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/session/SessionManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.time.Duration;
+import java.util.Collections;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/** Test for {@link SessionManager}. */
+public class SessionManagerTest {
```
Comment (reswqa) [inline flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/session/SessionManagerTest.java:43]: A small suggestion: I think we should now use Junit5 and assertj in newly introduced tests.

Comment (fsk119) [reply flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/session/SessionManagerTest.java]: Thanks for your suggestions. I use Junit5 in the current implementation.

**Review by wuchong — COMMENTED** [review]
I just left some minor comments. Others looks good.

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+        List<SqlGatewayEndpoint> endpoints = new ArrayList<>();
+        for (String identifier : identifiers) {
+            final SqlGatewayEndpointFactory factory =
+                    FactoryUtil.discoverFactory(
+                            Thread.currentThread().getContextClassLoader(),
+                            SqlGatewayEndpointFactory.class,
+                            identifier);
+            Configuration endpointConfig =
+                    new DelegatingConfiguration(
```
Comment (wuchong) [inline flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java:75]: Shall we need to delegate all configurations with the endpoint prefix? This disallows the endpoint to access session configurations, such as options in `SqlGatewayServiceConfigOptions` which might be helpful in some cases.

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+        @Override
+        public ReadableConfig getConfiguration() {
+            return configuration;
+        }
+
+        @Override
+        public Map<String, String> getOptions() {
+            return configuration.toMap();
+        }
```
Comment (wuchong) [inline flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java:137]: This looks like a design flaw that if `getConfiguration` and `getOptions` are equivalent, then why do we need to provide both of them? 

In DynamicTableFactory, the configuration represents the configs of the session, and options represent the table WITH options. I think we can change the methods to `getFlinkConfiguration()` and `getEndpointOptions()` to make it clear.

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java (hunk 2 above)
Comment (fsk119) [reply flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/endpoint/SqlGatewayEndpointFactoryUtils.java:78]: Yes. Currently we have `getFlinkConfiguration` and `getEndpointOptions`now. Users are able to get SqlGatewayService config from `getFlinkConfiguration.`

File: flink-table/flink-sql-gateway-api/src/test/java/org/apache/flink/table/gateway/api/utils/MockedSqlGatewayEndpointFactory.java [hunk 4]
```diff
@@ -41,12 +40,14 @@ public class MockedSqlGatewayEndpointFactory implements SqlGatewayEndpointFactor
...
-                .validate();
+        SqlGatewayEndpointFactoryUtils.EndpointFactoryHelper helper =
+                SqlGatewayEndpointFactoryUtils.createEndpointFactoryHelper(this, context);
+        helper.validate();
 
         return new MockedSqlGatewayEndpoint(
-                config.get(HOST), config.get(PORT), config.getOptional(DESCRIPTION).orElse(null));
+                helper.getOptions().get(HOST),
+                helper.getOptions().get(PORT),
+                helper.getOptions().getOptional(DESCRIPTION).orElse(null));
```
Comment (wuchong) [inline flink-table/flink-sql-gateway-api/src/test/java/org/apache/flink/table/gateway/api/utils/MockedSqlGatewayEndpointFactory.java:50]: minor: should be `helper.getOptions().get(DESCRIPTION)` to use default value if not present.

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by fsk119** [comment]
@flinkbot run azure

**Comment by fsk119** [comment]
@flinkbot run azure

**Pull request merged**